pkg/controller/volume/attachdetach/attach_detach_controller_test.go

```diff
@@ -81,6 +81,10 @@ func Test_AttachDetachControllerStateOfWorldPopulators_Positive(t *testing.T) {
 	fakeKubeClient := controllervolumetesting.CreateTestClient()
 	informerFactory := informers.NewSharedInformerFactory(fakeKubeClient, controller.NoResyncPeriodFunc())
 
+	var plugins []volume.VolumePlugin
+	plugins = append(plugins, controllervolumetesting.CreateTestPlugin(false)...)
+	plugins = append(plugins, csi.ProbeVolumePlugins()...)
```

Comment:
I'm thinking, do we really need to test CSI migration in  
`pkg/controller/volume/attachdetach` ? Maybe just the test plugin + a CSI PV instead of GCP PD would be enough to test the A/D controller, at least in most cases.  We could remove quite some code around CSINode and whatnot from the test framework here.

Would that be a way forward? Would there still be an unit tests somewhere else for in-tree translation to CSI + attach?

Reply:
Sorry for the delayed response.  @jsafrane @xing-yang 

>  I'm thinking, do we really need to test CSI migration in pkg/controller/volume/attachdetach? 

Answer 1:

Ideally, the ADC shouldn't be aware of plugin types, It does attach/detach operations through pre-defined interfaces and then updates the status of nodes.

However, in the current implementation, the ADC is aware of the csi migration, so we need to test it. If the csi plugin itself can support migrated other volume plugins, we can move this logic to the csi plugin and remove the csi migration logic from the ADC controller and elsewhere. 

It needs some refactoring work. we can create a new issue to discuss this if it has benefits. 

> Maybe just the test plugin + a CSI PV instead of GCP PD would be enough to test the A/D controller, at least in most cases. We could remove quite some code > around CSINode and whatnot from the test framework here.
> Would that be a way forward? Would there still be an unit tests somewhere else for in-tree translation to CSI + attach?

If `Answer 1` is implemented, the in-tree translation to CSI + attach test in ADC isn't needed. We can use the test plugin + a CSI PV to test the ADC controller. It should be enough to test the AD controller.

This PR aims to do some preparation work for https://github.com/kubernetes/kubernetes/pull/124023.

Reply:
Sounds good to me.  @jsafrane what do you think?

Reply:
kindly ping @jsafrane

Reply:
Ok, let's skip any huge A/D controller refactoring here.
/lgtm
/approve